<a href="https://colab.research.google.com/github/CervantesMCinthiaK/Metodos-Numericos/blob/main/Iterpolacion%20Neville%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Interpolación iterada de Neville
Alumna:Cinthia Karina Cervantes Martinez



In [19]:
#Definir la biblioteca que se usa
import numpy as np

In [20]:
# Solicitar datos al usuario
n = int(input("Ingrese el numero de nodos: "))

x = [] #almacena los nodos
f = [] #almacena su evaluacion es decir, f(x)

print("Ingrese los nodos (x) y sus valores correspondientes f(x):")
for i in range(n):
    xi = float(input(f"Ingrese x[{i + 1}]: "))
    fi = float(input(f"Ingrese f(x[{i + 1}]): "))
    x.append(xi)
    f.append(fi)

xp = float(input("Ingrese el valor de x donde desea evaluar P(x): "))

Ingrese el numero de nodos: 5
Ingrese los nodos (x) y sus valores correspondientes f(x):
Ingrese x[1]: 1.0
Ingrese f(x[1]): 0.7651977
Ingrese x[2]: 1.3
Ingrese f(x[2]): 0.6200860
Ingrese x[3]: 1.6
Ingrese f(x[3]): 0.4554022
Ingrese x[4]: 1.9
Ingrese f(x[4]): 0.2818186
Ingrese x[5]: 2.2
Ingrese f(x[5]): 0.110363
Ingrese el valor de x donde desea evaluar P(x): 1.5


In [21]:
# Validamos  para asegurar que los nodos sean unicos
if len(set(x)) != n:
    print("Los nodos deben ser unicos. Intente nuevamente.")
else:
    #PASO 1: Construccion de la tabla Qi,j
    Q = np.zeros((n, n))

    # Colocar los valores iniciales en la primera columna
    for i in range(n):
        Q[i, 0] = f[i]

    # Construimos  Qi,j
    for j in range(1, n):  # Columnas
        for i in range(j, n):  # Filas
            Q[i, j] = ((xp - x[i-j]) * Q[i, j-1] - (xp - x[i]) * Q[i-1, j-1]) / (x[i] - x[i-j]) #evaluamos en la formula dada en pag.91

    # Mostrar la tabla Q  Paso 2: Salida
    print("\nTabla Q:")
    encabezado = "i      xi           " + "   ".join([f"Q{i,j}" for j in range(n)])  #Crea una comprensión de lista que crea una lista de cadenas
    print(encabezado)

    for i in range(n):
        fila = [f"{Q[i, j]:.7f}" if i >= j else "         " for j in range(n)]
        print(f"{i:<6}{x[i]:<14.7f}" + "   ".join(fila)) #join une todos los elementos de la lista, # Imprime cada fila:i, x[i] y Q[i,j]

    # El valor del polinomio evaluado en xp
    Pxp = Q[n-1, n-1]
    print(f"\nP({xp}) = {Pxp:.7f}") #Aqui en el .7 se puede cambiar dependiendo cuantos decimales quiere.



Tabla Q:
i      xi           Q(4, 0)   Q(4, 1)   Q(4, 2)   Q(4, 3)   Q(4, 4)
0     1.0000000     0.7651977                                                
1     1.3000000     0.6200860   0.5233449                                    
2     1.6000000     0.4554022   0.5102968   0.5124715                        
3     1.9000000     0.2818186   0.5132634   0.5112857   0.5118127            
4     2.2000000     0.1103630   0.5104261   0.5137363   0.5118302   0.5118200

P(1.5) = 0.5118200


Se tiene un ejemplo del código realizado, es decir, usando Interpolación iterada de Neville con los datos \\
$x_0 = 1.0, \, f(x_0) = 0.7651977$ \\
$x_1 = 1.3, \, f(x_1) = 0.600860 $ \\
$x_2 = 1.6, \, f(x_2) = 0.4554022$ \\
$x_3 = 1.9, \, f(x_3) = 0.2818186$ \\
$x_4 = 2.2, \, f(x_4) = 0.1103623$ \\
hallamos el valor de $p(x)=1.5$